### PR TITLE
MIC should clean-up cache and any system files after run

### PR DESCRIPTION
**Pull request by Adub17030MS**
###### Merge Checklist  

**All** boxes should be checked before merging the PR *(just tick any boxes which don't apply to this PR)*
- [x] The toolchain has been rebuilt successfully (or no changes were made to it)
- [x] The toolchain/worker package manifests are up-to-date
- [x] Any updated packages successfully build (or no packages were changed)
- [x] Packages depending on static components modified in this PR (Golang, `*-static` subpackages, etc.) have had their `Release` tag incremented.
- [x] Package tests (%check section) have been verified with RUN_CHECK=y for existing SPEC files, or added to new SPEC files
- [x] All package sources are available
- [x] cgmanifest files are up-to-date and sorted (`./cgmanifest.json`, `./toolkit/scripts/toolchain/cgmanifest.json`, `.github/workflows/cgmanifest.json`)
- [x] LICENSE-MAP files are up-to-date (`./LICENSES-AND-NOTICES/SPECS/data/licenses.json`, `./LICENSES-AND-NOTICES/SPECS/LICENSES-MAP.md`, `./LICENSES-AND-NOTICES/SPECS/LICENSE-EXCEPTIONS.PHOTON`)
- [x] All source files have up-to-date hashes in the `*.signatures.json` files
- [x] `sudo make go-tidy-all` and `sudo make go-test-coverage` pass
- [x] Documentation has been updated to match any changes to the build system
- [x] Ready to merge

---

###### Summary 

When some commands are executed (ie: tdnf install), some files are populated in the cache or created locally. 
For MIC, we want to make sure that this is cleaned up before we snap the image. 

###### Change Log  

- Added `cleanUpImage.go` file, which has the functions that perform the image cleanup steps.
- We can add more functions to this file in the future for additional cleanup if needed.

###### Does this affect the toolchain?  

**NO**

###### Test Methodology

**Local build**

MIC image with no `cleanUpImage()` invocation:
image

MIC image with `cleanUpImage()` invocation:
image

Working unit test (Added more logging not included in this pr that shows the contents before the cleaup so we can see the unit test is properly verifying the cleanup occurs. When tdnf cache is cleaned up, the `offical-base` and `local-repo` folders remain and this is expected):
```
aditya@aditya-Virtual-Machine:~/azurelinux/toolkit/tools/pkg/imagecustomizerlib$ sudo go test -run TestCustomizeImagePackagesUpdate --base-image-core-efi-azl2 "$AZURE_LINUX_2_CORE_EFI_VHDX" -v
=== RUN   TestCustomizeImagePackagesUpdate
INFO[0000][main_test] Converting input image to a writeable format 
INFO[0000][main_test] Creating raw base image: /home/aditya/azurelinux/toolkit/tools/pkg/imagecustomizerlib/_tmp/TestCustomizeImagePackagesUpdate/build/image.raw 
INFO[0001][main_test] Image UUID: 1816036f-3fec-b837-de0b-cfe1c0f7513a 
INFO[0002][main_test] Overriding resolv.conf file                  
INFO[0022][main_test] Removing packages: [which]                   
INFO[0023][main_test] Updating base image packages                 
INFO[0035][main_test] Installing packages: [golang jq]             
INFO[0058][main_test] Updating packages: [jq]                      
INFO[0058][main_test] Contents of directory: /home/aditya/azurelinux/toolkit/tools/pkg/imagecustomizerlib/_tmp/TestCustomizeImagePackagesUpdate/build/imageroot/var/cache/tdnf 
INFO[0058][main_test]  - mariner-local-repo-02dc6fef               
INFO[0058][main_test]  - mariner-official-base                     
INFO[0058][main_test]  - mariner-official-base-1e9b3414            
INFO[0058][main_test]  - mariner-official-extras-be509c25          
INFO[0058][main_test]  - mariner-official-microsoft-125e652c       
INFO[0058][main_test] Cleaning up image                            
INFO[0058][main_test] Contents of directory: /home/aditya/azurelinux/toolkit/tools/pkg/imagecustomizerlib/_tmp/TestCustomizeImagePackagesUpdate/build/imageroot/var/cache/tdnf 
INFO[0058][main_test]  - mariner-local-repo-02dc6fef               
INFO[0058][main_test]  - mariner-official-base                     
INFO[0058][main_test] Creating image customizer release file       
INFO[0058][main_test] Restoring resolv.conf                        
INFO[0059][main_test] Adding systemd-resolved resolv.conf symlink  
INFO[0059][main_test] Checking for file system errors              
INFO[0059][main_test] Converting customized OS partitions into the final image 
INFO[0059][main_test] Writing: /home/aditya/azurelinux/toolkit/tools/pkg/imagecustomizerlib/_tmp/TestCustomizeImagePackagesUpdate/image.raw 
INFO[0060][main_test] Success!                                     
INFO[0062][main_test] Contents of directory: /home/aditya/azurelinux/toolkit/tools/pkg/imagecustomizerlib/_tmp/TestCustomizeImagePackagesUpdate/build/testimageroot/var/cache/tdnf 
INFO[0062][main_test]  - mariner-local-repo-02dc6fef               
INFO[0062][main_test]  - mariner-official-base                     
--- PASS: TestCustomizeImagePackagesUpdate (62.75s)
PASS
ok      github.com/microsoft/azurelinux/toolkit/tools/pkg/imagecustomizerlib    62.889s
```

### CODE REVIEW

File: toolkit/tools/pkg/imagecustomizerlib/customizeos.go [hunk 1]
```diff
@@ -29,6 +29,11 @@ func doOsCustomizations(buildDir string, baseConfigPath string, config *imagecus
 		return err
 	}
 
+	err = cleanUpImage(imageChroot)
```
Comment (romoh) [inline toolkit/tools/pkg/imagecustomizerlib/customizeos.go:32]: Thiis should be one of the last steps we take during customizations (before verity thou)

Comment (Adub17030MS) [reply toolkit/tools/pkg/imagecustomizerlib/customizeos.go]: Moved this step to before `enableOverlays()` and `enableVerityPartition()`

File: toolkit/tools/pkg/imagecustomizerlib/customizeos.go [hunk 2]
```diff
@@ -75,6 +75,11 @@ func doOsCustomizations(buildDir string, baseConfigPath string, config *imagecus
...
```
Comment (cwize1) [inline toolkit/tools/pkg/imagecustomizerlib/customizeos.go:78]: One of the policies of MIC is that we don’t do anything unless the user explicitly asks for something. As such any clean-up operations should be attached to the operations that made the leftover artifacts.

In this case, the tdnf cleanup code should be moved under `customizepackages.go` and should only be done if we installed or updated packages.

**Review by cwize1 — CHANGES_REQUESTED** [review]

File: toolkit/tools/pkg/imagecustomizerlib/cleanUpImage.go [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+		return shell.NewExecBuilder("tdnf", tdnfArgs...).
+			StdoutCallback(stdoutCallback).
+			LogLevel(shell.LogDisabledLevel, logrus.DebugLevel).
+			ErrorStderrLines(1).
+			Execute()
+	}
+
+	// Clear journal logs
+	clearJournalLogs := func() error {
+		return shell.NewExecBuilder("journalctl", "--vacuum-time=1s").
```
Comment (cwize1) [inline toolkit/tools/pkg/imagecustomizerlib/cleanUpImage.go:31]: With MIC, cleaning up the journald logs isn’t really necessary since the OS has never run and therefore has never had a change to produce logs.

The one case where this might be interesting is when customizing an image that has already been booted. This won’t be a common scenario. But it has come up at least once before.

However, this does touch on what I mentioned previously: MIC doesn’t make changes unless the user explicitly asks for them. So, any additional cleanup we do for the user will need to be attached to a new API. As such I recommend dropping the journald stuff from this PR.

File: toolkit/tools/pkg/imagecustomizerlib/customizeos.go (hunk 2 above)
Comment (romoh) [reply toolkit/tools/pkg/imagecustomizerlib/customizeos.go:78]: I see your point, but then a customer doing a tdnf install from a script will see a difference in image size and verity hash. Do we need to add explicit documentation on that behavior?

File: toolkit/tools/pkg/imagecustomizerlib/cleanUpImage.go [hunk 4]
```diff
@@ -0,0 +1,37 @@
+package imagecustomizerlib
+
+import (
+	"github.com/microsoft/azurelinux/toolkit/tools/internal/logger"
+	"github.com/microsoft/azurelinux/toolkit/tools/internal/safechroot"
+	"github.com/microsoft/azurelinux/toolkit/tools/internal/shell"
+	"github.com/sirupsen/logrus"
+)
+
+func cleanUpImage(imageChroot *safechroot.Chroot) error {
```
Comment (romoh) [inline toolkit/tools/pkg/imagecustomizerlib/cleanUpImage.go:10]: nit - should this be `cleanTdnfCache` then now?

File: toolkit/tools/pkg/imagecustomizerlib/cleanUpImage.go (hunk 3 above)
Comment (Adub17030MS) [reply toolkit/tools/pkg/imagecustomizerlib/cleanUpImage.go]: Got it, dropped the journald stuff from this PR.

File: toolkit/tools/pkg/imagecustomizerlib/cleanUpImage.go (hunk 4 above)
Comment (Adub17030MS) [reply toolkit/tools/pkg/imagecustomizerlib/cleanUpImage.go]: updated name to `cleanTdnfCache`

File: toolkit/tools/pkg/imagecustomizerlib/customizeos.go (hunk 2 above)
Comment (Adub17030MS) [reply toolkit/tools/pkg/imagecustomizerlib/customizeos.go]: @cwize1 thoughts?

File: toolkit/tools/pkg/imagecustomizerlib/cleanUpImage.go [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+
+import (
+	"github.com/microsoft/azurelinux/toolkit/tools/internal/logger"
+	"github.com/microsoft/azurelinux/toolkit/tools/internal/safechroot"
+	"github.com/microsoft/azurelinux/toolkit/tools/internal/shell"
+	"github.com/sirupsen/logrus"
+)
+
+func cleanTdnfCache(imageChroot *safechroot.Chroot) error {
+
```
Comment (cwize1) [inline toolkit/tools/pkg/imagecustomizerlib/cleanUpImage.go:11]: Don't add empty newlines after a `{`

File: toolkit/tools/pkg/imagecustomizerlib/cleanUpImage.go [hunk 6]
```diff
@@ -0,0 +1,37 @@
+package imagecustomizerlib
+
+import (
+	"github.com/microsoft/azurelinux/toolkit/tools/internal/logger"
+	"github.com/microsoft/azurelinux/toolkit/tools/internal/safechroot"
+	"github.com/microsoft/azurelinux/toolkit/tools/internal/shell"
+	"github.com/sirupsen/logrus"
+)
+
+func cleanTdnfCache(imageChroot *safechroot.Chroot) error {
```
Comment (cwize1) [inline toolkit/tools/pkg/imagecustomizerlib/cleanUpImage.go:10]: I'd probably move this function to `customizepackages.go` so it sits alongside the other tdnf helper functions.

File: toolkit/tools/pkg/imagecustomizerlib/cleanUpImage.go [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+
+	logger.Log.Infof("Cleaning up image")
+
+	// clear tdnf cache
+	tdnfClean := func() error {
+		tdnfArgs := []string{
+			"-v", "clean", "all",
+		}
+		stdoutCallback := func(line string) {
```
Comment (cwize1) [inline toolkit/tools/pkg/imagecustomizerlib/cleanUpImage.go:19]: Just use `LogLevel(logrus.TraceLevel, logrus.DebugLevel)`, You don't need to use `StdoutCallback` unless you want to do custom filtering.

File: toolkit/tools/pkg/imagecustomizerlib/cleanUpImage.go [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+	"github.com/microsoft/azurelinux/toolkit/tools/internal/shell"
+	"github.com/sirupsen/logrus"
+)
+
+func cleanTdnfCache(imageChroot *safechroot.Chroot) error {
+
+	logger.Log.Infof("Cleaning up image")
+
+	// clear tdnf cache
+	tdnfClean := func() error {
```
Comment (cwize1) [inline toolkit/tools/pkg/imagecustomizerlib/cleanUpImage.go:15]: Is the `tdnfClean` function really needed? Why not just put this code directly under `UnsafeRun()`?

File: toolkit/tools/pkg/imagecustomizerlib/customizeos.go (hunk 2 above)
Comment (cwize1) [reply toolkit/tools/pkg/imagecustomizerlib/customizeos.go]: If the user uses `tdnf` in a script, then they can call `tdnf clean all` as well.

**Review by cwize1 — CHANGES_REQUESTED** [review]

File: toolkit/tools/pkg/imagecustomizerlib/cleanUpImage.go (hunk 8 above)
Comment (Adub17030MS) [reply toolkit/tools/pkg/imagecustomizerlib/cleanUpImage.go:15]: Initially, in this PR, I was performing multiple clean-up steps, so I wrapped them each in their own functions. But since we are only performing the tdnf cleaning step now, I will put this code under `UnsafeRun()`

File: toolkit/tools/pkg/imagecustomizerlib/cleanUpImage.go (hunk 6 above)
Comment (Adub17030MS) [reply toolkit/tools/pkg/imagecustomizerlib/cleanUpImage.go:10]: Moved to `customizepackages.go`

File: toolkit/tools/pkg/imagecustomizerlib/cleanUpImage.go (hunk 5 above)
Comment (Adub17030MS) [reply toolkit/tools/pkg/imagecustomizerlib/cleanUpImage.go:11]: Updated

File: toolkit/tools/pkg/imagecustomizerlib/cleanUpImage.go (hunk 7 above)
Comment (Adub17030MS) [reply toolkit/tools/pkg/imagecustomizerlib/cleanUpImage.go]: Updated

File: toolkit/tools/pkg/imagecustomizerlib/cleanUpImage.go (hunk 8 above)
Comment (Adub17030MS) [reply toolkit/tools/pkg/imagecustomizerlib/cleanUpImage.go]: Updated and placed code under `UnsafeRun()`

**Review by romoh — APPROVED** [review]

File: toolkit/tools/pkg/imagecustomizerlib/customizepackages.go [hunk 9]
```diff
@@ -224,3 +229,21 @@ func isPackageInstalled(imageChroot *safechroot.Chroot, packageName string) bool
...
+	return imageChroot.UnsafeRun(func() error {
+		tdnfArgs := []string{
+			"-v", "clean", "all",
+		}
+		err := shell.NewExecBuilder("tdnf", tdnfArgs...).
+			LogLevel(logrus.TraceLevel, logrus.DebugLevel).
+			ErrorStderrLines(1).
+			Execute()
+		if err != nil {
+			return err
```
Comment (cwize1) [inline toolkit/tools/pkg/imagecustomizerlib/customizepackages.go:245]: You should use `fmt.Errorf` to tell the user what failed. This makes it easier to diagnose user reported errors.

File: toolkit/tools/pkg/imagecustomizerlib/customizepackages.go [hunk 10]
```diff
@@ -82,6 +82,11 @@ func addRemoveAndUpdatePackages(buildDir string, baseConfigPath string, config *
 		}
 	}
 
+	err = cleanTdnfCache(imageChroot)
```
Comment (cwize1) [inline toolkit/tools/pkg/imagecustomizerlib/customizepackages.go:85]: Place this code under a `if needRpmsSources {` check. Otherwise, it will always run.

Comment (cwize1) [inline toolkit/tools/pkg/imagecustomizerlib/customizepackages.go:85]: It would be good to update one (or more) of the `TestCustomizeImagePackages*` tests to verify the tdnf cache is in fact empty.

**Review by cwize1 — CHANGES_REQUESTED** [review]

File: toolkit/tools/pkg/imagecustomizerlib/customizepackages.go (hunk 10 above)
Comment (Adub17030MS) [reply toolkit/tools/pkg/imagecustomizerlib/customizepackages.go:85]: Updated the `TestCustomizeImagePackagesUpdate()` unit test to verify this as it is the most appropriate unit test here.

Comment (Adub17030MS) [reply toolkit/tools/pkg/imagecustomizerlib/customizepackages.go:85]: Done.

File: toolkit/tools/pkg/imagecustomizerlib/customizepackages.go (hunk 9 above)
Comment (Adub17030MS) [reply toolkit/tools/pkg/imagecustomizerlib/customizepackages.go:245]: Added!

File: toolkit/tools/pkg/imagecustomizerlib/customizepackages_test.go [hunk 11]
```diff
@@ -211,3 +217,54 @@ func TestCustomizeImagePackagesDiskSpace(t *testing.T) {
...
+
+func ensureTdnfCacheCleanup(t *testing.T, imageConnection *ImageConnection, dirPath string) error {
+
+	folderNames, err := logChrootDirectoryTest(imageConnection, "/var/cache/tdnf")
+	if err != nil {
+		return err
+	}
+
+	// Ensure there are exactly 2 folders after the tdnf cleanup
+	assert.Equal(t, 2, len(folderNames), "Expected exactly 2 folders, but got %d", len(folderNames))
```
Comment (cwize1) [inline toolkit/tools/pkg/imagecustomizerlib/customizepackages_test.go:252]: Are these two directories empty? If they are, if might be more interesting to count the total number of files under `/var/cache/tdnf` and make sure it is 0.

You can use [filepath.Walk](https://pkg.go.dev/path/filepath#Walk) to iterate over all the files and directories. (Which also might be useful for logging.)

**Review by cwize1 — CHANGES_REQUESTED** [review]

File: toolkit/tools/pkg/imagecustomizerlib/customizepackages_test.go [hunk 12]
```diff
@@ -211,3 +214,120 @@ func TestCustomizeImagePackagesDiskSpace(t *testing.T) {
 	assert.ErrorContains(t, err, "failed to customize raw image")
 	assert.ErrorContains(t, err, "failed to install package (gcc)")
 }
+
+func ensureTdnfCacheCleanup(t *testing.T, imageConnection *ImageConnection, dirPath string) error {
+	// Get version of base image
+	osVersion := getImageVersion(imageConnection)
+
+	// Variables to capture root directory contents
+	var rootDirContents []string
```
Comment (cwize1) [inline toolkit/tools/pkg/imagecustomizerlib/customizepackages_test.go:223]: I think the `ensureTdnfCacheCleanup` function could be greatly simplified by doing this instead:

1. Create a list of sub-directories in `/var/cache/tdnf` and the number of files within each sub-directory.
2. Verify all sub-directories are empty, except for `local-repo` on `2.0`.

File: toolkit/tools/pkg/imagecustomizerlib/customizepackages_test.go [hunk 13]
```diff
@@ -211,3 +214,120 @@ func TestCustomizeImagePackagesDiskSpace(t *testing.T) {
...
+	dirContents, err := searchChrootDirectory(imageConnection, dirPath)
+	if err != nil {
+		return err
+	}
+
+	// Filter and collect non-directory files, excluding those related to 'local-repo'
+	var existingFiles []string
+	for _, entry := range dirContents {
+		if !strings.Contains(entry, "local-repo") {
+			fileInfo, err := os.Stat(entry)
```
Comment (cwize1) [inline toolkit/tools/pkg/imagecustomizerlib/customizepackages_test.go:263]: If you changed `searchChrootDirectory` to only return files, then you wouldn't need to do this extra step here.

**Review by cwize1 — CHANGES_REQUESTED** [review]

File: toolkit/tools/pkg/imagecustomizerlib/customizepackages_test.go [hunk 14]
```diff
@@ -211,3 +215,37 @@ func TestCustomizeImagePackagesDiskSpace(t *testing.T) {
...
+			}
+			if !fileInfo.IsDir() {
+				// Append the file to the existingFiles array
+				existingFiles = append(existingFiles, path)
+			}
+		}
+		return nil
+	})
+
+	if err != nil {
```
Comment (cwize1) [inline toolkit/tools/pkg/imagecustomizerlib/customizepackages_test.go:243]: Use `assert.NoError(t, err)`

The caller of `ensureTdnfCacheCleanup` is currently ignoring the returned error.

**Review by cwize1 — CHANGES_REQUESTED** [review]

File: toolkit/tools/pkg/imagecustomizerlib/customizepackages_test.go [hunk 15]
```diff
@@ -211,3 +215,35 @@ func TestCustomizeImagePackagesDiskSpace(t *testing.T) {
 	assert.ErrorContains(t, err, "failed to customize raw image")
 	assert.ErrorContains(t, err, "failed to install package (gcc)")
 }
+
+func ensureTdnfCacheCleanup(t *testing.T, imageConnection *ImageConnection, dirPath string) error {
```
Comment (cwize1) [inline toolkit/tools/pkg/imagecustomizerlib/customizepackages_test.go:219]: You can probably remove the return error value.

**Review by cwize1 — APPROVED** [review]
